<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MFM: Multinomial Fine-Mapping • MFM</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MFM: Multinomial Fine-Mapping">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MFM</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MFM.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jennasimit/MFM">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MFM: Multinomial Fine-Mapping</h1>
                        <h4 class="author">Jenn Asimit</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:ja628@cam.ac.uk" class="email">ja628@cam.ac.uk</a>
      
                              <h4 class="author">Chris Wallace</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:cew54@cam.ac.uk" class="email">cew54@cam.ac.uk</a>
      
                  
            <h4 class="date">2019-02-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jennasimit/MFM/blob/master/vignettes/MFM.Rmd"><code>vignettes/MFM.Rmd</code></a></small>
      <div class="hidden name"><code>MFM.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>MFM is a package to simultaneously fine-map (select most likely set of causal variants) multiple related diseases with the same set of controls and share information between them. It relies on output from the package <a href="https://github.com/chr1swallace/GUESSFM">GUESSFM</a>, which fine-maps a single disease via stochastic search in a Bayesian framework using <a href="http://www.bgx.org.uk/software/guess.html">GUESS</a>.</p>
<p>This vignette introduces MFM, and gives an illustration with simulated data for two diseases with shared controls.</p>
</div>
<div id="mfm-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#mfm-overview" class="anchor"></a>MFM Overview</h1>
<p>The computational efficiency of MFM is greatly increased by making use of our approximation to the joint Bayes’ factor BF that depends only the individual disease BFs (from a logistic regression model) and the sample and model sizes. In the setting of <span class="math inline">\(d\)</span> diseases, the multinomial BF, <span class="math inline">\(B^M\)</span> is approximated by</p>
<p><span class="math display">\[B^M \approx \eta + \sum_d B_d^L, \]</span></p>
<p>where <span class="math inline">\(B_d^L\)</span> is the Bayes’ factor for disease <span class="math inline">\(d\)</span> and <span class="math inline">\(\eta\)</span> is a term that depends on the sample and model sizes.</p>
<p>Rather than reporting posterior probabilities for each SNP model, we focus on SNP groups that we construct such that SNPs in the same group are in LD, have a similar effect on disease, and rarely appear together in a disease model.</p>
</div>
<div id="simulation-example" class="section level1">
<h1 class="hasAnchor">
<a href="#simulation-example" class="anchor"></a>Simulation Example</h1>
<p>Genotype and phenotype data of two diseases with shared controls may be simulated using <a href="https://github.com/jennasimit/MFMextra">MFMextra</a>, together with <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">hapgen2</a> and reference panel data (e.g.  <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">CEU of 1000 Genomes</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(devtools)</a>
<a class="sourceLine" id="cb1-2" title="2">    <span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"jennasimit/MFMextra"</span>)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="co">#&gt; Skipping install of 'MFMextra' from a github remote, the SHA1 (56839e69) has not changed since last install.</span></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="co">#&gt;   Use `force = TRUE` to force installation</span></a></code></pre></div>
<p>First, we need to simulate some null data from which we will sample to generate the two sets of cases with shared controls. The entire genetic region of interest is simulated to maintain the linkage disequilibrium (LD) structure. In running simulations, there are typically some models of interest that are selected based on analysis of the data. For example, in a <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005272">previous fine-mapping</a> of <span class="math inline">\(IL2RA\)</span> in a large international sample, several SNP groups were identified as having the majority of the association signals with the autoimmune diseases multiple sclerosis (MS) and type 1 diabetes (T1D). These groups, together with previously identified lead SNPs for other autoimmune diseases (autoimmune thyroid disease (ATD; rs706799), alopecia areata (AA; rs3118470), rheumatoid arthritis (RA; rs10795791), and ulcerative colitis (UC; rs4147359)) will compose models that contribute to the non-negligible posterior probabilities. Therefore, for computational efficiency, we extract these SNPs from the generated data and focus on these in the fine-mapping simulation analysis.</p>
<p>Below is an example of <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">hapgen2</a> code to simulate the  region based on the <a href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">CEU of 1000 Genomes</a> reference panel, where keep-snps.txt would be the list of snp positions that are to be retained.</p>
<pre><code>    ./hapgen2
    -m ./genetic_map_chr10_combined_b37.txt \
    -l ./IL2RA.impute.legend \
    -h ./IL2RA.impute.hap \
    -n 100000 0 # 100,000 \
    -no_haps_output -no_gens_output \
    -t ./keep-snps.txt \
    -Ne 11418 \
    -o ./null_100k \</code></pre>
<p>Here is an example where diseases 1 and 2 each have two causal variants of which one is shared: both have causal variant rs61839660 in group A; disease 1 has additional causal variant rs56382813 in group D; disease 2 has additional causal variant rs11594656 in group C. The file “null_100k.controls.gen” is output from the above hapgen code and needed below. An example file is provided with this vignette, but for efficiency the convert.fn step has been run and the resulting matrix has been transposed for quicker reading into R (transposed matrix has SNPs as columns and individuals as rows).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MFMextra)</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="co">#&gt; Warning: replacing previous import 'data.table::melt' by 'reshape::melt'</span></a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co">#&gt; when loading 'GUESSFM'</span></a>
<a class="sourceLine" id="cb3-4" title="4">    <span class="co">#g0 &lt;- read.table(null_100k.controls.gen,header=FALSE,as.is=TRUE)</span></a>
<a class="sourceLine" id="cb3-5" title="5">    <span class="co">#Nn &lt;- (dim(g0)[2]-5)/3</span></a>
<a class="sourceLine" id="cb3-6" title="6">    <span class="co">#snpG &lt;- convert.fn(g0) # convert to a genotype matrix (snp rows, indiv cols)</span></a>
<a class="sourceLine" id="cb3-7" title="7">    </a>
<a class="sourceLine" id="cb3-8" title="8">    </a>
<a class="sourceLine" id="cb3-9" title="9">    dis &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"AD"</span>,<span class="st">"AC"</span>)</a>
<a class="sourceLine" id="cb3-10" title="10">    c12 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"rs61839660"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(snpG)) <span class="co"># A SNP for both diseases</span></a>
<a class="sourceLine" id="cb3-11" title="11">    c1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"rs56382813"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(snpG))  <span class="co"># D SNP for disease 1</span></a>
<a class="sourceLine" id="cb3-12" title="12">    c2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"rs11594656"</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(snpG))  <span class="co"># C SNP for disease 2</span></a>
<a class="sourceLine" id="cb3-13" title="13">    </a>
<a class="sourceLine" id="cb3-14" title="14">    causals1.ind &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(c12,c1)</a>
<a class="sourceLine" id="cb3-15" title="15">    causals2.ind &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(c12,c2)</a>
<a class="sourceLine" id="cb3-16" title="16">    prev &lt;-<span class="st"> </span><span class="fl">0.1</span> <span class="co"># prevalence for purpose of method evaluation</span></a>
<a class="sourceLine" id="cb3-17" title="17">    </a>
<a class="sourceLine" id="cb3-18" title="18">    N0 &lt;-<span class="st"> </span><span class="dv">3000</span> <span class="co"># disease 1 size</span></a>
<a class="sourceLine" id="cb3-19" title="19">    N1 &lt;-<span class="st"> </span><span class="dv">3000</span> <span class="co"># disease 2 size</span></a>
<a class="sourceLine" id="cb3-20" title="20">    N2 &lt;-<span class="st"> </span><span class="dv">3000</span> <span class="co"># controls size</span></a>
<a class="sourceLine" id="cb3-21" title="21">    ND=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">vector</a></span>(<span class="st">"list"</span>,<span class="dv">2</span>) <span class="co"># vector of sizes for cases</span></a>
<a class="sourceLine" id="cb3-22" title="22">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(ND) &lt;-<span class="st"> </span>dis</a>
<a class="sourceLine" id="cb3-23" title="23">    ND[[<span class="dv">1</span>]]&lt;-N1</a>
<a class="sourceLine" id="cb3-24" title="24">    ND[[<span class="dv">2</span>]]&lt;-N2</a>
<a class="sourceLine" id="cb3-25" title="25">    </a>
<a class="sourceLine" id="cb3-26" title="26">    </a>
<a class="sourceLine" id="cb3-27" title="27">    OR1a &lt;-<span class="st"> </span><span class="fl">1.4</span> <span class="co"># OR for A, disease 1</span></a>
<a class="sourceLine" id="cb3-28" title="28">    OR2a &lt;-<span class="st"> </span><span class="fl">1.25</span>  <span class="co"># OR for D, disease 1</span></a>
<a class="sourceLine" id="cb3-29" title="29">    OR1b &lt;-<span class="st"> </span><span class="fl">1.4</span> <span class="co"># OR for A, disease 2</span></a>
<a class="sourceLine" id="cb3-30" title="30">    OR2b &lt;-<span class="st"> </span><span class="fl">1.25</span>  <span class="co"># OR for C, disease 2</span></a>
<a class="sourceLine" id="cb3-31" title="31">    </a>
<a class="sourceLine" id="cb3-32" title="32"><span class="co"># snpG included in MFMextra package</span></a>
<a class="sourceLine" id="cb3-33" title="33">    sim &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/MFMextra/topics/phen.gen.fn">phen.gen.fn</a></span>(<span class="dt">beta1=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(prev),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR1a),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR2a)),<span class="dt">beta2=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(prev),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR1b),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(OR2b)),<span class="dt">snpG=</span>snpG,<span class="dt">N0=</span>N0,<span class="dt">N1=</span>N1,<span class="dt">N2=</span>N2,causals1.ind,causals2.ind)</a>
<a class="sourceLine" id="cb3-34" title="34">    Gm &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">"SnpMatrix"</span>,(sim<span class="op">$</span>G<span class="op">+</span><span class="dv">1</span>)) <span class="co"># snp cols, indivs rows # convert to SnpMatrix format, needed for GUESSFM</span></a>
<a class="sourceLine" id="cb3-35" title="35"><span class="co">#&gt; coercing object of mode  numeric  to SnpMatrix</span></a>
<a class="sourceLine" id="cb3-36" title="36">    Gm</a>
<a class="sourceLine" id="cb3-37" title="37"><span class="co">#&gt; A SnpMatrix with  9000 rows and  26 columns</span></a>
<a class="sourceLine" id="cb3-38" title="38"><span class="co">#&gt; Row names:  control.1 ... case2.3000 </span></a>
<a class="sourceLine" id="cb3-39" title="39"><span class="co">#&gt; Col names:  rs12722563 ... rs41295159</span></a></code></pre></div>
<p>Next, we run GUESSFM on each disease with controls and this is our input to MFM.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">    c0 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"control."</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(Gm))</a>
<a class="sourceLine" id="cb4-2" title="2">    c1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"case1."</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(Gm))</a>
<a class="sourceLine" id="cb4-3" title="3">    c2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"case2."</span>,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(Gm))</a>
<a class="sourceLine" id="cb4-4" title="4">    </a>
<a class="sourceLine" id="cb4-5" title="5">    G1 &lt;-<span class="st"> </span>Gm[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(c0,c1),] <span class="co"># SnpMatrix for disease 1 and controls</span></a>
<a class="sourceLine" id="cb4-6" title="6">    G2 &lt;-<span class="st"> </span>Gm[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(c0,c2),] <span class="co"># SnpMatrix for disease 2 and   controls</span></a>
<a class="sourceLine" id="cb4-7" title="7">    t1pheno &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">0</span>,N0),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span>,N1)) <span class="co"># phenotype vector for disease 1 and  controls</span></a>
<a class="sourceLine" id="cb4-8" title="8">    t2pheno &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">0</span>,N0),<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span>,N2)) <span class="co"># phenotype vector for disease 2 and  controls</span></a>
<a class="sourceLine" id="cb4-9" title="9">    </a>
<a class="sourceLine" id="cb4-10" title="10">    DIRin &lt;-<span class="st"> "tmpdirectory"</span> <span class="co"># path to a directory where to save GUESSFM results</span></a>
<a class="sourceLine" id="cb4-11" title="11">    mydir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(DIRin,<span class="st">"/GFMbvs"</span>,<span class="dt">sep=</span><span class="st">""</span>) <span class="co"># store GUESSFM results in a previously created directory DIRin</span></a>
<a class="sourceLine" id="cb4-12" title="12">    SM2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">vector</a></span>(<span class="st">"list"</span>,<span class="dv">2</span>) <span class="co"># collect the GUESSFM output from each disease</span></a>
<a class="sourceLine" id="cb4-13" title="13"></a>
<a class="sourceLine" id="cb4-14" title="14">      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(GUESSFM)      </a>
<a class="sourceLine" id="cb4-15" title="15">      <span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/run.bvs">run.bvs</a></span>(<span class="dt">X=</span>G1,<span class="dt">Y=</span>t1pheno,<span class="dt">tag.r2=</span>.<span class="dv">95</span>,<span class="dt">nexp=</span><span class="dv">3</span>,<span class="dt">nsave=</span><span class="dv">2000</span>,<span class="dt">gdir=</span>mydir,<span class="dt">wait=</span><span class="ot">TRUE</span>) <span class="co"># run with expected number of causal variants 3 for better mixing</span></a>
<a class="sourceLine" id="cb4-16" title="16"><span class="co">#&gt; max r2 is now0.942522800912621</span></a>
<a class="sourceLine" id="cb4-17" title="17"><span class="co">#&gt; saving tags object to tmpdirectory/GFMbvs/tags.RData</span></a>
<a class="sourceLine" id="cb4-18" title="18"><span class="co">#&gt; using 6000 samples 18 SNPs, 0 covariates.</span></a>
<a class="sourceLine" id="cb4-19" title="19"><span class="co">#&gt; rs2104286    MAF=0.31    p=1.8728e-15</span></a>
<a class="sourceLine" id="cb4-20" title="20"><span class="co">#&gt; rs56382813   MAF=0.25    p=0.00023829</span></a>
<a class="sourceLine" id="cb4-21" title="21"><span class="co">#&gt; rs12722508   MAF=0.15    p=2.4288e-05</span></a>
<a class="sourceLine" id="cb4-22" title="22"><span class="co">#&gt; setting prior parameters nexp, sd: 3 1.58113883008419</span></a>
<a class="sourceLine" id="cb4-23" title="23"><span class="co">#&gt; creating decode files under tmpdirectory/GFMbvs</span></a>
<a class="sourceLine" id="cb4-24" title="24"><span class="co">#&gt; conditional analysis suggests a starting model with 0-based columns: 2 6 14</span></a>
<a class="sourceLine" id="cb4-25" title="25"><span class="co">#&gt; aka: rs2104286 rs56382813 rs12722508</span></a>
<a class="sourceLine" id="cb4-26" title="26"><span class="co">#&gt; creating input files under tmpdirectory/GFMbvs</span></a>
<a class="sourceLine" id="cb4-27" title="27"><span class="co">#&gt; writing to tmpdirectory/GFMbvs/init_55000:</span></a>
<a class="sourceLine" id="cb4-28" title="28"><span class="co">#&gt; 3</span></a>
<a class="sourceLine" id="cb4-29" title="29"><span class="co">#&gt; 2</span></a>
<a class="sourceLine" id="cb4-30" title="30"><span class="co">#&gt; 6</span></a>
<a class="sourceLine" id="cb4-31" title="31"><span class="co">#&gt; 14</span></a>
<a class="sourceLine" id="cb4-32" title="32"><span class="co">#&gt; running GUESS with command</span></a>
<a class="sourceLine" id="cb4-33" title="33"><span class="co">#&gt; /Library/Frameworks/R.framework/Versions/3.5/Resources/library/R2GUESS/bin//GUESS -history -X tmpdirectory/GFMbvs/X_55000 -Y tmpdirectory/GFMbvs/Y_55000 -nsweep 55000 -burn_in 5000 -out tmpdirectory/GFMbvs/out_55000 -par tmpdirectory/GFMbvs/par.xml -top 2000 -init tmpdirectory/GFMbvs/init_55000 -Egam 3 -Sgam 1.58113883008419 -n_chain 3 &gt; tmpdirectory/GFMbvs/log</span></a>
<a class="sourceLine" id="cb4-34" title="34"><span class="co">#&gt; [1] "/Library/Frameworks/R.framework/Versions/3.5/Resources/library/R2GUESS/bin//GUESS -history -X tmpdirectory/GFMbvs/X_55000 -Y tmpdirectory/GFMbvs/Y_55000 -nsweep 55000 -burn_in 5000 -out tmpdirectory/GFMbvs/out_55000 -par tmpdirectory/GFMbvs/par.xml -top 2000 -init tmpdirectory/GFMbvs/init_55000 -Egam 3 -Sgam 1.58113883008419 -n_chain 3 &gt; tmpdirectory/GFMbvs/log"</span></a>
<a class="sourceLine" id="cb4-35" title="35">      d &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/read.snpmod">read.snpmod</a></span>(mydir)</a>
<a class="sourceLine" id="cb4-36" title="36"><span class="co">#&gt; reading filestub: tmpdirectory/GFMbvs/decode_55000</span></a>
<a class="sourceLine" id="cb4-37" title="37"><span class="co">#&gt; reading filestub: tmpdirectory/GFMbvs/out_55000</span></a>
<a class="sourceLine" id="cb4-38" title="38"><span class="co">#&gt; reading from tmpdirectory/GFMbvs/out_55000_output_best_visited_models.txt</span></a>
<a class="sourceLine" id="cb4-39" title="39"><span class="co">#&gt; 387 models, with pp &gt; 7.93e-19, and, cumulatively, to 1</span></a>
<a class="sourceLine" id="cb4-40" title="40">      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/load">load</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(mydir,<span class="st">"tags.RData"</span>))</a>
<a class="sourceLine" id="cb4-41" title="41">      dx &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/expand.tags">expand.tags</a></span>(d,tags)</a>
<a class="sourceLine" id="cb4-42" title="42"><span class="co">#&gt; expanding tags for 387 models over 18 tag SNPs, tagging a total of 26 SNPs.</span></a>
<a class="sourceLine" id="cb4-43" title="43">      best &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/best.models">best.models</a></span>(dx,<span class="dt">pp.thr=</span><span class="fl">0.0001</span>)</a>
<a class="sourceLine" id="cb4-44" title="44">      abf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/abf.calc">abf.calc</a></span>(<span class="dt">y=</span>t1pheno,<span class="dt">x=</span>G1,<span class="dt">models=</span>best<span class="op">$</span>str,<span class="dt">family=</span><span class="st">"binomial"</span>)</a>
<a class="sourceLine" id="cb4-45" title="45"><span class="co">#&gt; Calculating BICs using speedglm</span></a>
<a class="sourceLine" id="cb4-46" title="46"><span class="co">#&gt; 171 of 171 models can be evaluated directly</span></a>
<a class="sourceLine" id="cb4-47" title="47"><span class="co">#&gt; calculating ABF</span></a>
<a class="sourceLine" id="cb4-48" title="48"><span class="co">#&gt; evaluating direct models, n=171</span></a>
<a class="sourceLine" id="cb4-49" title="49">      SM2[[<span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/abf2snpmod">abf2snpmod</a></span>(abf,<span class="dt">expected=</span><span class="dv">2</span>,<span class="dt">nsnps=</span><span class="dv">854</span>) <span class="co"># find approximate Bayes' factors (ABFs) using expected number of causal variants 2 and the number of SNPs in the re$</span></a>
<a class="sourceLine" id="cb4-50" title="50"><span class="co">#&gt; creating snpmod data.frame</span></a>
<a class="sourceLine" id="cb4-51" title="51"><span class="co">#&gt; calculating marginal SNP inclusion probabilities</span></a>
<a class="sourceLine" id="cb4-52" title="52">    </a>
<a class="sourceLine" id="cb4-53" title="53">      <span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/run.bvs">run.bvs</a></span>(<span class="dt">X=</span>G2,<span class="dt">Y=</span>t2pheno,<span class="dt">tag.r2=</span>.<span class="dv">95</span>,<span class="dt">nexp=</span><span class="dv">3</span>,<span class="dt">nsave=</span><span class="dv">2000</span>,<span class="dt">gdir=</span>mydir,<span class="dt">wait=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-54" title="54"><span class="co">#&gt; max r2 is now0.931034178357753</span></a>
<a class="sourceLine" id="cb4-55" title="55"><span class="co">#&gt; saving tags object to tmpdirectory/GFMbvs/tags.RData</span></a>
<a class="sourceLine" id="cb4-56" title="56"><span class="co">#&gt; using 6000 samples 18 SNPs, 0 covariates.</span></a>
<a class="sourceLine" id="cb4-57" title="57"><span class="co">#&gt; rs706779 MAF=0.47    p=&lt; 2.22e-16</span></a>
<a class="sourceLine" id="cb4-58" title="58"><span class="co">#&gt; rs35285258   MAF=0.24    p=5.5668e-05</span></a>
<a class="sourceLine" id="cb4-59" title="59"><span class="co">#&gt; rs61839660   MAF=0.13    p=0.00028614</span></a>
<a class="sourceLine" id="cb4-60" title="60"><span class="co">#&gt; setting prior parameters nexp, sd: 3 1.58113883008419</span></a>
<a class="sourceLine" id="cb4-61" title="61"><span class="co">#&gt; creating decode files under tmpdirectory/GFMbvs</span></a>
<a class="sourceLine" id="cb4-62" title="62"><span class="co">#&gt; conditional analysis suggests a starting model with 0-based columns: 3 5 9</span></a>
<a class="sourceLine" id="cb4-63" title="63"><span class="co">#&gt; aka: rs706779 rs35285258 rs61839660</span></a>
<a class="sourceLine" id="cb4-64" title="64"><span class="co">#&gt; creating input files under tmpdirectory/GFMbvs</span></a>
<a class="sourceLine" id="cb4-65" title="65"><span class="co">#&gt; writing to tmpdirectory/GFMbvs/init_55000:</span></a>
<a class="sourceLine" id="cb4-66" title="66"><span class="co">#&gt; 3</span></a>
<a class="sourceLine" id="cb4-67" title="67"><span class="co">#&gt; 3</span></a>
<a class="sourceLine" id="cb4-68" title="68"><span class="co">#&gt; 5</span></a>
<a class="sourceLine" id="cb4-69" title="69"><span class="co">#&gt; 9</span></a>
<a class="sourceLine" id="cb4-70" title="70"><span class="co">#&gt; running GUESS with command</span></a>
<a class="sourceLine" id="cb4-71" title="71"><span class="co">#&gt; /Library/Frameworks/R.framework/Versions/3.5/Resources/library/R2GUESS/bin//GUESS -history -X tmpdirectory/GFMbvs/X_55000 -Y tmpdirectory/GFMbvs/Y_55000 -nsweep 55000 -burn_in 5000 -out tmpdirectory/GFMbvs/out_55000 -par tmpdirectory/GFMbvs/par.xml -top 2000 -init tmpdirectory/GFMbvs/init_55000 -Egam 3 -Sgam 1.58113883008419 -n_chain 3 &gt; tmpdirectory/GFMbvs/log</span></a>
<a class="sourceLine" id="cb4-72" title="72"><span class="co">#&gt; [1] "/Library/Frameworks/R.framework/Versions/3.5/Resources/library/R2GUESS/bin//GUESS -history -X tmpdirectory/GFMbvs/X_55000 -Y tmpdirectory/GFMbvs/Y_55000 -nsweep 55000 -burn_in 5000 -out tmpdirectory/GFMbvs/out_55000 -par tmpdirectory/GFMbvs/par.xml -top 2000 -init tmpdirectory/GFMbvs/init_55000 -Egam 3 -Sgam 1.58113883008419 -n_chain 3 &gt; tmpdirectory/GFMbvs/log"</span></a>
<a class="sourceLine" id="cb4-73" title="73">      d &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/read.snpmod">read.snpmod</a></span>(mydir)</a>
<a class="sourceLine" id="cb4-74" title="74"><span class="co">#&gt; reading filestub: tmpdirectory/GFMbvs/decode_55000</span></a>
<a class="sourceLine" id="cb4-75" title="75"><span class="co">#&gt; reading filestub: tmpdirectory/GFMbvs/out_55000</span></a>
<a class="sourceLine" id="cb4-76" title="76"><span class="co">#&gt; reading from tmpdirectory/GFMbvs/out_55000_output_best_visited_models.txt</span></a>
<a class="sourceLine" id="cb4-77" title="77"><span class="co">#&gt; 369 models, with pp &gt; 5.33e-24, and, cumulatively, to 1</span></a>
<a class="sourceLine" id="cb4-78" title="78">      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/load">load</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(mydir,<span class="st">"tags.RData"</span>))</a>
<a class="sourceLine" id="cb4-79" title="79">      dx &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/expand.tags">expand.tags</a></span>(d,tags)</a>
<a class="sourceLine" id="cb4-80" title="80"><span class="co">#&gt; expanding tags for 369 models over 18 tag SNPs, tagging a total of 26 SNPs.</span></a>
<a class="sourceLine" id="cb4-81" title="81">      best &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/best.models">best.models</a></span>(dx,<span class="dt">pp.thr=</span><span class="fl">0.0001</span>)</a>
<a class="sourceLine" id="cb4-82" title="82">      abf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/abf.calc">abf.calc</a></span>(<span class="dt">y=</span>t2pheno,<span class="dt">x=</span>G2,<span class="dt">models=</span>best<span class="op">$</span>str,<span class="dt">family=</span><span class="st">"binomial"</span>)</a>
<a class="sourceLine" id="cb4-83" title="83"><span class="co">#&gt; Calculating BICs using speedglm</span></a>
<a class="sourceLine" id="cb4-84" title="84"><span class="co">#&gt; 298 of 298 models can be evaluated directly</span></a>
<a class="sourceLine" id="cb4-85" title="85"><span class="co">#&gt; calculating ABF</span></a>
<a class="sourceLine" id="cb4-86" title="86"><span class="co">#&gt; evaluating direct models, n=298</span></a>
<a class="sourceLine" id="cb4-87" title="87">      SM2[[<span class="dv">2</span>]] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/GUESSFM/topics/abf2snpmod">abf2snpmod</a></span>(abf,<span class="dt">expected=</span><span class="dv">2</span>,<span class="dt">nsnps=</span><span class="dv">854</span>)</a>
<a class="sourceLine" id="cb4-88" title="88"><span class="co">#&gt; creating snpmod data.frame</span></a>
<a class="sourceLine" id="cb4-89" title="89"><span class="co">#&gt; calculating marginal SNP inclusion probabilities</span></a>
<a class="sourceLine" id="cb4-90" title="90">    </a>
<a class="sourceLine" id="cb4-91" title="91">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(SM2) &lt;-<span class="st"> </span>dis</a>
<a class="sourceLine" id="cb4-92" title="92"></a>
<a class="sourceLine" id="cb4-93" title="93">    target.odds &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># could also provide a vector here; this setting corresponds to a 50:50 odds of non-sharing to sharing of causal variants between diseases</span></a>
<a class="sourceLine" id="cb4-94" title="94">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MFM)</a>
<a class="sourceLine" id="cb4-95" title="95">    <span class="co"># snpGroups included in MFMextra package</span></a>
<a class="sourceLine" id="cb4-96" title="96">    PP &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PPmarginal.multiple.fn.html">PPmarginal.multiple.fn</a></span>(SM2,dis,<span class="dt">thr=</span><span class="fl">0.999</span>,target.odds,N0,ND,<span class="dt">nsnps=</span><span class="dv">854</span>)</a>
<a class="sourceLine" id="cb4-97" title="97"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb4-98" title="98"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb4-99" title="99"><span class="co">#&gt; CPP threshold = 0.999</span></a>
<a class="sourceLine" id="cb4-100" title="100"><span class="co">#&gt;  n.each (AD/AC) = 41/42</span></a></code></pre></div>
<p>The PP object is a list with components: * PP, containing the SNP model posterior probabilities for each disease and each target odds * MPP, containing the SNP marginal posterior probabilities of inclusion and each target odds</p>
<p>It is easier to interpret in terms of SNP groups, so next run the following, where the snpGroups object is available in the MFMextra package:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">    mpp.pp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MPP.PP.groups.fn.html">MPP.PP.groups.fn</a></span>(PP<span class="op">$</span>MPP,PP<span class="op">$</span>PP,dis,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"null"</span>,target.odds),snpGroups)</a>
<a class="sourceLine" id="cb5-2" title="2">mpp.pp</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co">#&gt; $mppGS</span></a>
<a class="sourceLine" id="cb5-4" title="4"><span class="co">#&gt;                A B            C D E F            J          sAA</span></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="co">#&gt; AD.null 1.000000 0 0.0000000000 1 0 0 0.000000e+00 0.000000e+00</span></a>
<a class="sourceLine" id="cb5-6" title="6"><span class="co">#&gt; AD.1    1.000026 0 0.0003405729 1 0 0 7.889971e-05 3.452468e-05</span></a>
<a class="sourceLine" id="cb5-7" title="7"><span class="co">#&gt; AC.null 1.000000 0 1.0000000000 0 0 0 0.000000e+00 0.000000e+00</span></a>
<a class="sourceLine" id="cb5-8" title="8"><span class="co">#&gt; AC.1    0.996590 0 0.9970383813 0 0 0 4.170396e-03 2.107452e-04</span></a>
<a class="sourceLine" id="cb5-9" title="9"><span class="co">#&gt;                  sUC          sRA</span></a>
<a class="sourceLine" id="cb5-10" title="10"><span class="co">#&gt; AD.null 0.000000e+00 0.000000e+00</span></a>
<a class="sourceLine" id="cb5-11" title="11"><span class="co">#&gt; AD.1    2.596052e-05 2.700061e-05</span></a>
<a class="sourceLine" id="cb5-12" title="12"><span class="co">#&gt; AC.null 0.000000e+00 0.000000e+00</span></a>
<a class="sourceLine" id="cb5-13" title="13"><span class="co">#&gt; AC.1    0.000000e+00 1.717710e-05</span></a>
<a class="sourceLine" id="cb5-14" title="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-15" title="15"><span class="co">#&gt; $gPP</span></a>
<a class="sourceLine" id="cb5-16" title="16"><span class="co">#&gt; $gPP[[1]]</span></a>
<a class="sourceLine" id="cb5-17" title="17"><span class="co">#&gt;         null            1</span></a>
<a class="sourceLine" id="cb5-18" title="18"><span class="co">#&gt; null       0 1.463500e-14</span></a>
<a class="sourceLine" id="cb5-19" title="19"><span class="co">#&gt; A.D        1 9.994667e-01</span></a>
<a class="sourceLine" id="cb5-20" title="20"><span class="co">#&gt; A.D.sAA    0 3.452468e-05</span></a>
<a class="sourceLine" id="cb5-21" title="21"><span class="co">#&gt; A.D.sUC    0 2.596052e-05</span></a>
<a class="sourceLine" id="cb5-22" title="22"><span class="co">#&gt; A.D.sRA    0 2.700061e-05</span></a>
<a class="sourceLine" id="cb5-23" title="23"><span class="co">#&gt; A.C.D      0 3.405729e-04</span></a>
<a class="sourceLine" id="cb5-24" title="24"><span class="co">#&gt; A.D.J      0 7.889971e-05</span></a>
<a class="sourceLine" id="cb5-25" title="25"><span class="co">#&gt; A.A.D      0 2.632000e-05</span></a>
<a class="sourceLine" id="cb5-26" title="26"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-27" title="27"><span class="co">#&gt; $gPP[[2]]</span></a>
<a class="sourceLine" id="cb5-28" title="28"><span class="co">#&gt;         null            1</span></a>
<a class="sourceLine" id="cb5-29" title="29"><span class="co">#&gt; null       0 1.016979e-19</span></a>
<a class="sourceLine" id="cb5-30" title="30"><span class="co">#&gt; A.C        1 9.956017e-01</span></a>
<a class="sourceLine" id="cb5-31" title="31"><span class="co">#&gt; J          0 2.961619e-03</span></a>
<a class="sourceLine" id="cb5-32" title="32"><span class="co">#&gt; C.J        0 4.483924e-04</span></a>
<a class="sourceLine" id="cb5-33" title="33"><span class="co">#&gt; A.C.J      0 7.603852e-04</span></a>
<a class="sourceLine" id="cb5-34" title="34"><span class="co">#&gt; A.C.sAA    0 2.107452e-04</span></a>
<a class="sourceLine" id="cb5-35" title="35"><span class="co">#&gt; A.C.sRA    0 1.717710e-05</span></a></code></pre></div>
<p>The mpp.pp object is a list with componets:</p>
<ul>
<li><p>mppGS: matrix of SNP group MPP</p></li>
<li><p>gPP: list of disease SNP group PP matrices</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#mfm-overview">MFM Overview</a></li>
      <li><a href="#simulation-example">Simulation Example</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jenn Asimit, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
